### PR TITLE
[reconfiguration] support adding validators

### PR DESCRIPTION
**Pull request by sblackshear**
- Added the `additions` vector to the `LibraSystem.ValidatorSet` resource. This vector tracks the list of addresses to be added to the validator set at the time of the next epoch.
- Added `LibraSystem.add_validator(addr)` procedure that can only be invoked by the Association. This procedure will cause the validator whose `ValidatorConfig` is stored at `addr` to be added to the validator set at the next epoch change.
- Modified `LibraSysterm.reconfigure()` to add each address in the `additions` vector to the validator set.
- Simplified genesis generation code to simply populate the `additions` vector and rely on `reconfigure` to create the validator set.
- E2E tests for addition flow, authentication of `add_validator`, error cases of `add_validator`, interaction with key rotation

## Motivation

Support for 1 of the 2 remaining reconfiguration features. Removal to follow.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Lots of new E2E tests. Also tested by the genesis tx generation code

### CODE REVIEW

File: language/stdlib/modules/libra_system.mvir [hunk 1]
```diff
@@ -236,14 +257,31 @@ module LibraSystem {
...
+           network_identity_pubkey: ValidatorConfig.network_identity_pubkey(&config),
+       };
+    }
+
     // Trigger a reconfiguation the Libra system by:
-    // (1) Computing a new validator set and storing it on chain
-    // (2) Emitting an event containing new validator set, which will be passed to the executor
-    // TODO: make  this private + only callable from block prologue
+    // (1) Checking if there have been any additions since the last reconfiguration (adding if so)
+    // (2) Checking if there have been any
```
Comment (zekun000) [inline language/stdlib/modules/libra_system.mvir]: any key rotation?

File: language/functional_tests/tests/testsuite/validator_set/add_validator.mvir [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// check: ABORTED
+// check: 18
+
+//! new-transaction
+//! sender: alice
+import 0x0.ValidatorConfig;
+main() {
+    ValidatorConfig.register_candidate_validator(h"10", h"20", h"30");
```
Comment (zekun000) [inline language/functional_tests/tests/testsuite/validator_set/add_validator.mvir]: anyone can register as candidate?

File: language/stdlib/modules/libra_system.mvir (hunk 1 above)
Comment (sblackshear) [reply language/stdlib/modules/libra_system.mvir]: Yes, oops! Will fix.

**Comment by sblackshear** [comment]
> is the test-lib.bpl file auto generated? if so we probably want to mark it to avoid review

Good point; I'll check with @shazqadeer and the Move prover folks to identify all of the autogenerated files and mark them in a follow-up.

File: language/functional_tests/tests/testsuite/validator_set/add_validator.mvir (hunk 2 above)
Comment (sblackshear) [reply language/functional_tests/tests/testsuite/validator_set/add_validator.mvir]: Yes. Registering as a candidate is necessary, but not sufficient to become a validator. The Association account needs to call `add_validator` on an address that's already registered.

**Comment by shazqadeer** [comment]
All the files with .bpl extension are auto-generated by the Boogie tool.

File: language/stdlib/modules/libra_system.mvir [hunk 3]
```diff
@@ -253,12 +297,36 @@ module LibraSystem {
         // TODO: support adding and removing validators. Eventually, we will do this by computing
         // the new validator set from a larger list of candidate validators sorted by stake.
         validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        additions_ref = &mut copy(validator_set_ref).additions;
         vector_ref = &mut copy(validator_set_ref).validators;
 
+        // (1) Check for additions
+        i = 0;
+        len = Vector.length<address>(freeze(copy(additions_ref)));
+        if (copy(len) > 0) {
```
Comment (tzakian) [inline language/stdlib/modules/libra_system.mvir]: What do you think about making the dependency of `changed` on `len` here more explicit? e.g. 

```rust 
changed = copy(len) > 0;
if (copy(changed)) {
   ....
}
```

Comment (sblackshear) [reply language/stdlib/modules/libra_system.mvir]: I will add a comment for this.

**Pull request closed without merging**